README.md
# RP2040_SD Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/RP2040_SD.svg?)](https://www.ardu-badge.com/RP2040_SD)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/RP2040_SD.svg)](https://github.com/khoih-prog/RP2040_SD/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/RP2040_SD/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/RP2040_SD.svg)](http://github.com/khoih-prog/RP2040_SD/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>


---
---

## Table of Contents

* [Why do we need this RP2040_SD library](#why-do-we-need-this-rp2040_sd-library)
  * [Features](#features)
  * [Currently supported Boards](#currently-supported-boards)
* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [Examples](#examples)
  * [  1. CardInfo](examples/CardInfo)
  * [  2. DataLogger](examples/DataLogger)
  * [  3. DumpFile](examples/DumpFile)
  * [  4. Files](examples/Files)
  * [  5. ListFiles](examples/ListFiles)
  * [  6. NonBlockingWrite](examples/NonBlockingWrite)
  * [  7. ReadWrite](examples/ReadWrite)
* [Example ReadWrite](#example-readwrite)
  * [ 1. File ReadWrite.ino](#1-file-readwriteino)
* [Debug Terminal Output Samples](#debug-terminal-output-samples)
  * [ 1. CardInfo on MBED RaspberryPi Pico](#1-cardinfo-on-mbed-raspberrypi-pico)
  * [ 2. DumpFile on MBED RaspberryPi Pico](#2-dumpfile-on-mbed-raspberrypi-pico)
  * [ 3. ListFiles on MBED RaspberryPi Pico](#3-listfiles-on-mbed-raspberrypi-pico)
  * [ 4. ReadWrite on MBED RaspberryPi Pico](#4-readwrite-on-mbed-raspberrypi-pico)
  * [ 5. Files on MBED RaspberryPi Pico](#5-files-on-mbed-raspberrypi-pico)
  * [ 6. CardInfo on RASPBERRY_PI_PICO](#6-cardinfo-on-raspberry_pi_pico)
  * [ 7. DumpFile on RASPBERRY_PI_PICO](#7-dumpfile-on-raspberry_pi_pico)
  * [ 8. ListFiles on RASPBERRY_PI_PICO](#8-listfiles-on-raspberry_pi_pico)
  * [ 9. ReadWrite on RASPBERRY_PI_PICO](#9-readwrite-on-raspberry_pi_pico)
  * [10. Files on RASPBERRY_PI_PICO](#10-files-on-raspberry_pi_pico)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [Contributions and Thanks](#contributions-and-thanks)
* [Contributing](#contributing)
* [License](#license)
* [Copyright](#copyright)

---
---

### Why do we need this [RP2040_SD library](https://github.com/khoih-prog/RP2040_SD)

## Features

This is an Arduino library for **RP2040-based SPI SD Cards**.

It's based on and modified from :

1. [Bill Greiman's SdFat library](https://github.com/greiman/SdFat)
2. [Adafruit SD library](https://github.com/adafruit/SD)

Have a look at [SdFat's README](https://github.com/greiman/SdFat/blob/master/README.md)

```
SdFat Version 2 supports FAT16/FAT32 and exFAT SD cards. It is mostly backward compatible with SdFat Version 1 for FAT16/FAT32 cards.

exFAT supports files larger than 4GB so files sizes and positions are type uint64_t for classes that support exFAT.

exFAT has many features not available in FAT16/FAT32.  exFAT has excellent support for contiguous files on flash devices and supports preallocation.

If the SD card is the only SPI device, use dedicated SPI mode. This can greatly improve performance.
```

This library was created as an effort to use SD Card while the `arduino-pico` core still has issue [**SD card FILE_WRITE issue** #214](https://github.com/earlephilhower/arduino-pico/issues/214), which has been fixed from core v1.8.6.

It's better to use the built-in [SD library](https://github.com/earlephilhower/arduino-pico/tree/master/libraries/SD) for [**Earle Philhower's arduino-pico core v1.8.6+**](https://github.com/earlephilhower/arduino-pico). 

But if you need to write codes to be used in both [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed) and [**Earle Philhower's arduino-pico core**](https://github.com/earlephilhower/arduino-pico) without having to modify the code, you can have the option to use this library.

---

### Currently supported Boards

1. RP2040-based boards such as **NANO_RP2040_CONNECT, RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, etc. using either [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed) or [**Earle Philhower's arduino-pico core**](https://github.com/earlephilhower/arduino-pico).


---
---

## Prerequisites

1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
2. [`Arduino mbed_rp2040 core 3.4.1+`](https://github.com/arduino/ArduinoCore-mbed) for Arduino (Use Arduino Board Manager) RP2040-based boards, such as **Arduino Nano RP2040 Connect, RASPBERRY_PI_PICO, etc.**. [![GitHub release](https://img.shields.io/github/release/arduino/ArduinoCore-mbed.svg)](https://github.com/arduino/ArduinoCore-mbed/releases/latest)
3. [`Earle Philhower's arduino-pico core v2.6.3+`](https://github.com/earlephilhower/arduino-pico) for RP2040-based boards such as **RASPBERRY_PI_PICO, ADAFRUIT_FEATHER_RP2040 and GENERIC_RP2040**, etc. [![GitHub release](https://img.shields.io/github/release/earlephilhower/arduino-pico.svg)](https://github.com/earlephilhower/arduino-pico/releases/latest)

---
---

## Installation

### Use Arduino Library Manager

The best and easiest way is to use `Arduino Library Manager`. Search for [**RP2040_SD**](https://github.com/khoih-prog/RP2040_SD), then select / install the latest version.
You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/RP2040_SD.svg?)](https://www.ardu-badge.com/RP2040_SD) for more detailed instructions.

### Manual Install

Another way to install is to:

1. Navigate to [**RP2040_SD**](https://github.com/khoih-prog/RP2040_SD) page.
2. Download the latest release `RP2040_SD-main.zip`.
3. Extract the zip file to `RP2040_SD-main` directory 
4. Copy whole `RP2040_SD-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**RP2040_SD** library](https://registry.platformio.org/libraries/khoih-prog/RP2040_SD) by using [Library Manager](https://registry.platformio.org/libraries/khoih-prog/RP2040_SD/installation). Search for **RP2040_SD** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Khoi%20Hoang%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)


---
---

### Examples 

 1. [CardInfo](examples/CardInfo)
 2. [DataLogger](examples/DataLogger)
 3. [DumpFile](examples/DumpFile)
 4. [Files](examples/Files)
 5. [ListFiles](examples/ListFiles)
 6. [NonBlockingWrite](examples/NonBlockingWrite)
 7. [ReadWrite](examples/ReadWrite)


---
---

### Example [ReadWrite](examples/ReadWrite)

#### 1. File [ReadWrite.ino](examples/ReadWrite/ReadWrite.ino)

https://github.com/khoih-prog/RP2040_SD/blob/c1c62cc86fcfac3f90614888e4aa41c409c5505e/examples/ReadWrite/ReadWrite.ino#L26-L167

---
---

### Debug Terminal Output Samples

#### 1. CardInfo on MBED RaspberryPi Pico

The following is the sample terminal output when running example [CardInfo](examples/CardInfo) on MBED RaspberryPi Pico using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed)

```cpp
Starting SD Card CardInfo on MBED RaspberryPi Pico
MBED RP2040_SD v1.0.1
Initializing SD card with SS = 5
SCK = 2
MOSI = 3
MISO = 4
Wiring is correct and a card is present.

Card type:         SDHC
Clusters:          982896
Blocks x Cluster:  64
Total Blocks:      62905344

Volume type is:    FAT32
Volume size (Kb):  31452672
Volume size (Mb):  30715
Volume size (Gb):  30.00
Initialization done.

Files found on the card (name, date and size in bytes): 
DATALOG.TXT		144
TEST.TXT		36
NEWTEST.TXT		36
NEWTEST1.TXT		18
MYDATA~1.TXT		0
NEWTEST2.TXT		54
NEWTEST3.TXT		54
NEWTEST4.TXT		72
NEWTEST0.TXT		297
NEWDATA.TXT		18
```

---

#### 2. DumpFile on MBED RaspberryPi Pico

The following is the sample terminal output when running example [DumpFile](examples/DumpFile) on MBED RaspberryPi Pico using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed)

```cpp
Starting SD Card DumpFile on MBED RaspberryPi Pico
MBED RP2040_SD v1.0.1
Initializing SD card with SS = 5
SCK = 2
MOSI = 3
MISO = 4
Initialization done.
====== Data ======
Testing RP2040 SD
====== Data ======
[SD] Closed " newdata.txt ", nfilecount =  0
```

---

#### 3. ListFiles on MBED RaspberryPi Pico

The following is the sample terminal output when running example [ListFiles](examples/ListFiles) on MBED RaspberryPi Pico using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed)

```cpp
Starting SD Card ListFiles on MBED RaspberryPi Pico
MBED RP2040_SD v1.0.1
Initializing SD card with SS = 5
SCK = 2
MOSI = 3
MISO = 4
Initialization done.
DATALOG.TXT		126
TEST.TXT		36
NEWTEST.TXT		36
NEWTEST1.TXT		18
MYDATA~1.TXT		0
NEWTEST2.TXT		54
NEWTEST3.TXT		54
NEWTEST4.TXT		72
NEWTEST0.TXT		297
Print Directory done!
writeData = Testing RP2040 SD
readData = Testing RP2040 SD
DATALOG.TXT		144
TEST.TXT		36
NEWTEST.TXT		36
NEWTEST1.TXT		18
MYDATA~1.TXT		0
NEWTEST2.TXT		54
NEWTEST3.TXT		54
NEWTEST4.TXT		72
NEWTEST0.TXT		297
NEWDATA.TXT		18
Print Directory done!
```

---

#### 4. ReadWrite on MBED RaspberryPi Pico

The following is the sample terminal output when running example [ReadWrite](examples/ReadWrite) on MBED RaspberryPi Pico using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed)


```cpp
Starting SD Card ReadWrite on MBED RaspberryPi Pico
MBED RP2040_SD v1.0.1
Initializing SD card with SS = 5
SCK = 2
MOSI = 3
MISO = 4
initialization done.
Writing to newtest0.txt ==> Testing writing to newtest0.txt
[SD] Closed " newtest0.txt ", nfilecount =  0
done.
Reading from newtest0.txt
===============
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
[SD] Closed " newtest0.txt ", nfilecount =  0
===============
```

---

#### 5. Files on MBED RaspberryPi Pico

The following is the sample terminal output when running example [Files](examples/Files) on MBED RaspberryPi Pico using [**Arduino-mbed RP2040** core](https://github.com/arduino/ArduinoCore-mbed)


```cpp
Starting SD Card Files on MBED RaspberryPi Pico
MBED RP2040_SD v1.0.1
Initializing SD card with SS = 5
SCK = 2
MOSI = 3
MISO = 4
Initialization done.
example.txt doesn't exist.
Creating example.txt
writeData = 0xDEADBEEF
[SD] Closed " example.txt ", nfilecount =  0
readData = 0xDEADBEEF
[SD] Closed " example.txt ", nfilecount =  0
example.txt exists.
Removing example.txt...
example.txt doesn't exist.
```

---
---

#### 6. CardInfo on RASPBERRY_PI_PICO

The following is the sample terminal output when running example [CardInfo](examples/CardInfo) on RASPBERRY_PI_PICO using [**arduino-pico core**](https://github.com/earlephilhower/arduino-pico)

```cpp
Starting SD Card CardInfo on RASPBERRY_PI_PICO
RP2040_SD v1.0.1
Initializing SD card with SS = 17
SCK = 18
MOSI = 19
MISO = 16
Wiring is correct and a card is present.

Card type:         SDHC
Clusters:          1965597
Blocks x Cluster:  32
Total Blocks:      62899104

Volume type is:    FAT32
Volume size (Kb):  31449552
Volume size (Mb):  30712
Volume size (Gb):  29.99
Initialization done.

Files found on the card (name, date and size in bytes): 
DATALOG.TXT		267
DEMO.TXT		360960
CANADA~1.JPG		11156
TEST.TXT		72
NEWTEST4.TXT		132
NEWTEST0.TXT		330
NEWDATA.TXT		54
```

---

#### 7. DumpFile on RASPBERRY_PI_PICO

The following is the sample terminal output when running example [DumpFile](examples/DumpFile) on RASPBERRY_PI_PICO using [**arduino-pico core**](https://github.com/earlephilhower/arduino-pico)


```cpp
Starting SD Card DumpFile on RASPBERRY_PI_PICO
RP2040_SD v1.0.1
Initializing SD card with SS = 17
SCK = 18
MOSI = 19
MISO = 16
Initialization done.
====== Data ======
Testing RP2040 SD
====== Data ======
[SD] Closed " newdata.txt ", nfilecount =  0
```

---

#### 8. ListFiles on RASPBERRY_PI_PICO

The following is the sample terminal output when running example [ListFiles](examples/ListFiles) on RASPBERRY_PI_PICO using [**arduino-pico core**](https://github.com/earlephilhower/arduino-pico)

```cpp
Starting SD Card ListFiles on RASPBERRY_PI_PICO
RP2040_SD v1.0.1
Initializing SD card with SS = 17
SCK = 18
MOSI = 19
MISO = 16
Initialization done.
DATALOG.TXT		249
DEMO.TXT		82432
CANADA~1.JPG		11156
TEST.TXT		72
NEWTEST4.TXT		132
NEWTEST0.TXT		330
NEWDATA.TXT		36
Print Directory done!
writeData = Testing RP2040 SD
readData = Testing RP2040 SD
DATALOG.TXT		267
DEMO.TXT		82432
CANADA~1.JPG		11156
TEST.TXT		72
NEWTEST4.TXT		132
NEWTEST0.TXT		330
NEWDATA.TXT		54
Print Directory done!
```

---

#### 9. ReadWrite on RASPBERRY_PI_PICO

The following is the sample terminal output when running example [ReadWrite](examples/ReadWrite) on RASPBERRY_PI_PICO using [**arduino-pico core**](https://github.com/earlephilhower/arduino-pico)

```cpp
Starting SD Card ReadWrite on RASPBERRY_PI_PICO
RP2040_SD v1.0.1
Initializing SD card with SS = 17
SCK = 18
MOSI = 19
MISO = 16
initialization done.
Writing to newtest0.txt ==> Testing writing to newtest0.txt
[SD] Closed " newtest0.txt ", nfilecount =  0
done.
Reading from newtest0.txt
===============
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt
Testing writing to newtest0.txt

changelog.md
# RP2040_SD Library

[![arduino-library-badge](https://www.ardu-badge.com/badge/RP2040_SD.svg?)](https://www.ardu-badge.com/RP2040_SD)
[![GitHub release](https://img.shields.io/github/release/khoih-prog/RP2040_SD.svg)](https://github.com/khoih-prog/RP2040_SD/releases)
[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/khoih-prog/RP2040_SD/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg?style=flat)](#Contributing)
[![GitHub issues](https://img.shields.io/github/issues/khoih-prog/RP2040_SD.svg)](http://github.com/khoih-prog/RP2040_SD/issues)


<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Donate to my libraries using BuyMeACoffee" style="height: 50px !important;width: 181px !important;" ></a>
<a href="https://www.buymeacoffee.com/khoihprog6" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v1.0.1](#releases-v101)
  * [Releases v1.0.0](#releases-v100)

---
---


## Changelog

### Releases v1.0.1

1. Fix platform in `library.json`

### Releases v1.0.0

1. Port to RP2040-based boards using Arduino-mbed or arduino-pico core
2. Add debugging feature




src/utility/FatStructs.h
/****************************************************************************************************************************
  FatStructs.h
  
  For all RP2040 boads using Arduimo-mbed or arduino-pico core
  
  RP2040_SD is a library enable the usage of SD on RP2040-based boards
  
  This Library is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

  This Library is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

  You should have received a copy of the GNU General Public License along with the Arduino SdFat Library.  
  If not, see <http://www.gnu.org/licenses/>.
  
  Based on and modified from  Arduino SdFat Library (https://github.com/arduino/Arduino)
  
  (C) Copyright 2009 by William Greiman
  (C) Copyright 2010 SparkFun Electronics
  (C) Copyright 2021 by Khoi Hoang
  
  Built by Khoi Hoang https://github.com/khoih-prog/RP2040_SD
  Licensed under GPL-3.0 license
  
  Version: 1.0.1
  
  Version Modified By   Date      Comments
  ------- -----------  ---------- -----------
  1.0.0  K Hoang       18/06/2021 Port to RP2040-based boards using Arduimo-mbed or arduino-pico core
  1.0.1  K Hoang       22/10/2021 Fix platform in library.json for PIO
 *****************************************************************************************************************************/

#pragma once

#ifndef FatStructs_h
#define FatStructs_h

/**
   \file
   FAT file structures
*/
/*
   mostly from Microsoft document fatgen103.doc
   http://www.microsoft.com/whdc/system/platform/firmware/fatgen.mspx
*/
//------------------------------------------------------------------------------
/** Value for byte 510 of boot block or MBR */
#define BOOTSIG0      0x55

/** Value for byte 511 of boot block or MBR */
#define BOOTSIG1      0xAA
//------------------------------------------------------------------------------
/**
   \struct partitionTable
   \brief MBR partition table entry

   A partition table entry for a MBR formatted storage device.
   The MBR partition table has four entries.
*/
struct partitionTable
{
  /**
     Boot Indicator . Indicates whether the volume is the active
     partition.  Legal values include: 0X00. Do not use for booting.
     0X80 Active partition.
  */
  uint8_t  boot;
  /**
      Head part of Cylinder-head-sector address of the first block in
      the partition. Legal values are 0-255. Only used in old PC BIOS.
  */
  uint8_t  beginHead;
  /**
     Sector part of Cylinder-head-sector address of the first block in
     the partition. Legal values are 1-63. Only used in old PC BIOS.
  */
  unsigned beginSector : 6;
  /** High bits cylinder for first block in partition. */
  unsigned beginCylinderHigh : 2;
  /**
     Combine beginCylinderLow with beginCylinderHigh. Legal values
     are 0-1023.  Only used in old PC BIOS.
  */
  uint8_t  beginCylinderLow;
  /**
     Partition type. See defines that begin with PART_TYPE_ for
     some Microsoft partition types.
  */
  uint8_t  type;
  /**
     head part of cylinder-head-sector address of the last sector in the
     partition.  Legal values are 0-255. Only used in old PC BIOS.
  */
  uint8_t  endHead;
  /**
     Sector part of cylinder-head-sector address of the last sector in
     the partition.  Legal values are 1-63. Only used in old PC BIOS.
  */
  unsigned endSector : 6;
  /** High bits of end cylinder */
  unsigned endCylinderHigh : 2;
  /**
     Combine endCylinderLow with endCylinderHigh. Legal values
     are 0-1023.  Only used in old PC BIOS.
  */
  uint8_t  endCylinderLow;
  /** Logical block address of the first block in the partition. */
  uint32_t firstSector;
  /** Length of the partition, in blocks. */
  uint32_t totalSectors;
} __attribute__((packed));

/** Type name for partitionTable */
typedef struct partitionTable part_t;
//------------------------------------------------------------------------------
/**
   \struct masterBootRecord

   \brief Master Boot Record

   The first block of a storage device that is formatted with a MBR.
*/
struct masterBootRecord
{
  /** Code Area for master boot program. */
  uint8_t  codeArea[440];
  /** Optional WindowsNT disk signature. May contain more boot code. */
  uint32_t diskSignature;
  /** Usually zero but may be more boot code. */
  uint16_t usuallyZero;
  /** Partition tables. */
  part_t   part[4];
  /** First MBR signature byte. Must be 0X55 */
  uint8_t  mbrSig0;
  /** Second MBR signature byte. Must be 0XAA */
  uint8_t  mbrSig1;
} __attribute__((packed));

/** Type name for masterBootRecord */
typedef struct masterBootRecord mbr_t;

//------------------------------------------------------------------------------
/**
   \struct biosParmBlock

   \brief BIOS parameter block

    The BIOS parameter block describes the physical layout of a FAT volume.
*/
struct biosParmBlock
{
  /**
     Count of bytes per sector. This value may take on only the
     following values: 512, 1024, 2048 or 4096
  */
  uint16_t bytesPerSector;
  /**
     Number of sectors per allocation unit. This value must be a
     power of 2 that is greater than 0. The legal values are
     1, 2, 4, 8, 16, 32, 64, and 128.
  */
  uint8_t  sectorsPerCluster;
  /**
     Number of sectors before the first FAT.
     This value must not be zero.
  */
  uint16_t reservedSectorCount;
  /** The count of FAT data structures on the volume. This field should
      always contain the value 2 for any FAT volume of any type.
  */
  uint8_t  fatCount;
  /**
    For FAT12 and FAT16 volumes, this field contains the count of
    32-byte directory entries in the root directory. For FAT32 volumes,
    this field must be set to 0. For FAT12 and FAT16 volumes, this
    value should always specify a count that when multiplied by 32
    results in a multiple of bytesPerSector.  FAT16 volumes should
    use the value 512.
  */
  uint16_t rootDirEntryCount;
  /**
     This field is the old 16-bit total count of sectors on the volume.
     This count includes the count of all sectors in all four regions
     of the volume. This field can be 0; if it is 0, then totalSectors32
     must be non-zero.  For FAT32 volumes, this field must be 0. For
     FAT12 and FAT16 volumes, this field contains the sector count, and
     totalSectors32 is 0 if the total sector count fits
     (is less than 0x10000).
  */
  uint16_t totalSectors16;
  /**
     This dates back to the old MS-DOS 1.x media determination and is
     no longer usually used for anything.  0xF8 is the standard value
     for fixed (non-removable) media. For removable media, 0xF0 is
     frequently used. Legal values are 0xF0 or 0xF8-0xFF.
  */
  uint8_t  mediaType;
  /**
     Count of sectors occupied by one FAT on FAT12/FAT16 volumes.
     On FAT32 volumes this field must be 0, and sectorsPerFat32
     contains the FAT size count.
  */
  uint16_t sectorsPerFat16;
  /** Sectors per track for interrupt 0x13. Not used otherwise. */
  uint16_t sectorsPerTrtack;
  /** Number of heads for interrupt 0x13.  Not used otherwise. */
  uint16_t headCount;
  /**
     Count of hidden sectors preceding the partition that contains this
     FAT volume. This field is generally only relevant for media
      visible on interrupt 0x13.
  */
  uint32_t hidddenSectors;
  /**
     This field is the new 32-bit total count of sectors on the volume.
     This count includes the count of all sectors in all four regions
     of the volume.  This field can be 0; if it is 0, then
     totalSectors16 must be non-zero.
  */
  uint32_t totalSectors32;
  /**
     Count of sectors occupied by one FAT on FAT32 volumes.
  */
  uint32_t sectorsPerFat32;
  /**
     This field is only defined for FAT32 media and does not exist on
     FAT12 and FAT16 media.
     Bits 0-3 -- Zero-based number of active FAT.
                 Only valid if mirroring is disabled.
     Bits 4-6 -- Reserved.
     Bit 7  -- 0 means the FAT is mirrored at runtime into all FATs.
            -- 1 means only one FAT is active; it is the one referenced in bits 0-3.
     Bits 8-15  -- Reserved.
  */
  uint16_t fat32Flags;
  /**
     FAT32 version. High byte is major revision number.
     Low byte is minor revision number. Only 0.0 define.
  */
  uint16_t fat32Version;
  /**
     Cluster number of the first cluster of the root directory for FAT32.
     This usually 2 but not required to be 2.
  */
  uint32_t fat32RootCluster;
  /**
     Sector number of FSINFO structure in the reserved area of the
     FAT32 volume. Usually 1.
  */
  uint16_t fat32FSInfo;
  /**
     If non-zero, indicates the sector number in the reserved area
     of the volume of a copy of the boot record. Usually 6.
     No value other than 6 is recommended.
  */
  uint16_t fat32BackBootBlock;
  /**
     Reserved for future expansion. Code that formats FAT32 volumes
     should always set all of the bytes of this field to 0.
  */
  uint8_t  fat32Reserved[12];
} __attribute__((packed));

/** Type name for biosParmBlock */
typedef struct biosParmBlock bpb_t;

//------------------------------------------------------------------------------
/**
   \struct fat32BootSector

   \brief Boot sector for a FAT16 or FAT32 volume.

*/
struct fat32BootSector
{
  /** X86 jmp to boot program */
  uint8_t  jmpToBootCode[3];
  /** informational only - don't depend on it */
  char     oemName[8];
  /** BIOS Parameter Block */
  bpb_t    bpb;
  /** for int0x13 use value 0X80 for hard drive */
  uint8_t  driveNumber;
  /** used by Windows NT - should be zero for FAT */
  uint8_t  reserved1;
  /** 0X29 if next three fields are valid */
  uint8_t  bootSignature;
  /** usually generated by combining date and time */
  uint32_t volumeSerialNumber;
  /** should match volume label in root dir */
  char     volumeLabel[11];
  /** informational only - don't depend on it */
  char     fileSystemType[8];
  /** X86 boot code */
  uint8_t  bootCode[420];
  /** must be 0X55 */
  uint8_t  bootSectorSig0;
  /** must be 0XAA */
  uint8_t  bootSectorSig1;
} __attribute__((packed));

//------------------------------------------------------------------------------
// End Of Chain values for FAT entries
/** FAT16 end of chain value used by Microsoft. */
#define FAT16EOC        (uint16_t) 0XFFFF

/** Minimum value for FAT16 EOC.  Use to test for EOC. */
#define FAT16EOC_MIN    (uint16_t) 0XFFF8

/** FAT32 end of chain value used by Microsoft. */
#define FAT32EOC        (uint32_t) 0x0FFFFFFF

/** Minimum value for FAT32 EOC.  Use to test for EOC. */
#define FAT32EOC_MIN    (uint32_t) 0x0FFFFFF8

/** Mask a for FAT32 entry. Entries are 28 bits. */
#define FAT32MASK       (uint32_t) 0x0FFFFFFF

/** Type name for fat32BootSector */
typedef struct fat32BootSector fbs_t;

//------------------------------------------------------------------------------
/**
   \struct directoryEntry
   \brief FAT short directory entry

   Short means short 8.3 name, not the entry size.

   Date Format. A FAT directory entry date stamp is a 16-bit field that is
   basically a date relative to the MS-DOS epoch of 01/01/1980. Here is the
   format (bit 0 is the LSB of the 16-bit word, bit 15 is the MSB of the
   16-bit word):

   Bits 9-15: Count of years from 1980, valid value range 0-127
   inclusive (1980-2107).

   Bits 5-8: Month of year, 1 = January, valid value range 1-12 inclusive.

   Bits 0-4: Day of month, valid value range 1-31 inclusive.

   Time Format. A FAT directory entry time stamp is a 16-bit field that has
   a granularity of 2 seconds. Here is the format (bit 0 is the LSB of the
   16-bit word, bit 15 is the MSB of the 16-bit word).

   Bits 11-15: Hours, valid value range 0-23 inclusive.

   Bits 5-10: Minutes, valid value range 0-59 inclusive.

   Bits 0-4: 2-second count, valid value range 0-29 inclusive (0 - 58 seconds).

   The valid time range is from Midnight 00:00:00 to 23:59:58.
*/
struct directoryEntry
{
  /**
     Short 8.3 name.
     The first eight bytes contain the file name with blank fill.
     The last three bytes contain the file extension with blank fill.
  */
  uint8_t  name[11];
  /** Entry attributes.

     The upper two bits of the attribute byte are reserved and should
     always be set to 0 when a file is created and never modified or
     looked at after that.  See defines that begin with DIR_ATT_.
  */
  uint8_t  attributes;
  /**
     Reserved for use by Windows NT. Set value to 0 when a file is
     created and never modify or look at it after that.
  */
  uint8_t  reservedNT;
  /**
     The granularity of the seconds part of creationTime is 2 seconds
     so this field is a count of tenths of a second and its valid
     value range is 0-199 inclusive. (WHG note - seems to be hundredths)
  */
  uint8_t  creationTimeTenths;
  /** Time file was created. */
  uint16_t creationTime;
  /** Date file was created. */
  uint16_t creationDate;
  /**
     Last access date. Note that there is no last access time, only
     a date.  This is the date of last read or write. In the case of
     a write, this should be set to the same date as lastWriteDate.
  */
  uint16_t lastAccessDate;
  /**
     High word of this entry's first cluster number (always 0 for a
     FAT12 or FAT16 volume).
  */
  uint16_t firstClusterHigh;
  /** Time of last write. File creation is considered a write. */
  uint16_t lastWriteTime;
  /** Date of last write. File creation is considered a write. */
  uint16_t lastWriteDate;
  /** Low word of this entry's first cluster number. */
  uint16_t firstClusterLow;
  /** 32-bit unsigned holding this file's size in bytes. */
  uint32_t fileSize;
};    // __attribute__((packed));

//------------------------------------------------------------------------------
// Definitions for directory entries
//
/** Type name for directoryEntry */
typedef struct directoryEntry dir_t;

enum
{
  DIR_NAME_0XE5           = 0X05,
  DIR_NAME_DELETED        = 0XE5,
  DIR_NAME_FREE           = 0X00,
  DIR_ATT_READ_ONLY       = 0X01,
  DIR_ATT_HIDDEN          = 0X02,
  DIR_ATT_SYSTEM          = 0X04,
  DIR_ATT_VOLUME_ID       = 0X08,
  DIR_ATT_DIRECTORY       = 0X10,
  DIR_ATT_ARCHIVE         = 0X20,
  DIR_ATT_LONG_NAME       = 0X0F, 
};

/** Test mask for long name entry */
#define DIR_ATT_LONG_NAME_MASK    0X3F

/** defined attribute bits */
#define DIR_ATT_DEFINED_BITS      0X3F

/** Directory entry is part of a long name */
static inline uint8_t DIR_IS_LONG_NAME(const dir_t* dir)
{
  return (dir->attributes & DIR_ATT_LONG_NAME_MASK) == DIR_ATT_LONG_NAME;
}

/** Mask for file/subdirectory tests */
uint8_t const DIR_ATT_FILE_TYPE_MASK = (DIR_ATT_VOLUME_ID | DIR_ATT_DIRECTORY);

/** Directory entry is for a file */
static inline uint8_t DIR_IS_FILE(const dir_t* dir)
{
  return (dir->attributes & DIR_ATT_FILE_TYPE_MASK) == 0;
}

/** Directory entry is for a subdirectory */
static inline uint8_t DIR_IS_SUBDIR(const dir_t* dir)
{
  return (dir->attributes & DIR_ATT_FILE_TYPE_MASK) == DIR_ATT_DIRECTORY;
}

/** Directory entry is for a file or subdirectory */
static inline uint8_t DIR_IS_FILE_OR_SUBDIR(const dir_t* dir)
{
  return (dir->attributes & DIR_ATT_VOLUME_ID) == 0;
}

#endif  // FatStructs_h
